File: linux_cd_cmd_simulation/cpp/mycd.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

/**
 * This function checks whether character is alphanumeric or not
 * @param ch character to be validated
 * @return Return true if character is alphanumeric else false
 */
bool is_alnum(char ch) {
    // Check whether character is alphanumeric or not.
    return !isalnum(ch);
}

/**
 * This function validates dirctory name
 * @param dir_name Directory name
 * @return Return true if dir_name has only alphanumeric characters else false
 */
bool validate_dir_path_name(string dir_name) {
    if (dir_name.compare("..") == 0 || dir_name.compare(".") == 0) {
        return true;
    }

    return find_if(dir_name.begin(), dir_name.end(), is_alnum) == dir_name.end();
}

/**
 * This function is used to get directory name from directory path and validates it.
 * @param dir_path Directory path
 * @param index Index is used as starting point in dir_path string to get next dirctory name in path
 * @param separator Separator is used to separate directory name
 * @return Returns directory name
 */
 string get_and_validate_directory_name(string dir_path, int &index, char separator) {
    // Check whether index is greater than or equal to the length of dir_path
    if(index >= dir_path.length()) {
        return dir_path;
    }
    string res = "";

    for(; index < dir_path.length(); index++) {
        if(dir_path[index] == '/') {
            break;
        } else {
            res.push_back(dir_path[index]);
        }
    }
    bool is_alphanum = validate_dir_path_name(res);

    if(is_alphanum) {
        return res;
    } else {
        string expt = ": No such file or directory: ";
        expt += res;
        throw expt;
    }
}

/**
 * This function is used to extract all directory names from directory path
 * @param dir_path Directory path (/abc/edf)
 * @return Returns list of direcrory names corresponding to directory path. 
 */
vector<string> convert_dir_path_to_list(string dir_path) {
    vector<string> res;
    string dir_name = "";

    for(int index = 0; index < dir_path.length(); ) {
        if(dir_path[index] == '/') {
            res.push_back("/");
            index++;
            while(index < dir_path.length() && dir_path[index] == '/') {
                index++;
            }
        } else {
            char separator = '/';
            dir_name = get_and_validate_directory_name(dir_path, index, separator);
            res.push_back(dir_name);
        }
    }
    return res;
}

/**
 * This function is used to prepare new dirctory path
 * @param cwd_dir_path Current working directory
 * @param new_dir_path Destination directory path
 * @return Returns new directory path
 */
vector<string> prepare_new_dir(vector<string> cwd_dir_path, vector<string> new_dir_path) {

	if (new_dir_path[0].compare("/") == 0) {
        cwd_dir_path.clear();
        cwd_dir_path.push_back("/");
	}

	for(int i = 0; i < new_dir_path.size(); i++) {
		if (new_dir_path[i].compare(".") == 0 || new_dir_path[i].compare("/") == 0) {
			continue;
		}
		if (new_dir_path[i].compare("..") == 0) {
			if (cwd_dir_path.size() > 1) {
				cwd_dir_path.pop_back();
                if(cwd_dir_path[cwd_dir_path.size() - 1] == "/") {
                    cwd_dir_path.pop_back();
                }
			}
		} else {
            if (cwd_dir_path[cwd_dir_path.size() - 1] != "/") {
                cwd_dir_path.push_back("/");
            }
            cwd_dir_path.push_back(new_dir_path[i]);
		}
    }
    if(cwd_dir_path.size() == 0) {
        cwd_dir_path.push_back("/");
    }
	return cwd_dir_path;
}

int main(int argc, char **argv) {
    try {
        // Validate number of passed parameters from terminal
        if(argc < 3 || argc > 3) {
            if(argc < 3) {
                cout<<"mycd: too few arguments";
            } else {
                cout<<"mycd: too many arguments";
            }
            return 1;
        }

	string cwd = argv[1];  // Current working directory path
	string to_dirctory = argv[2];  // To new directory path

	vector<string> cwd_dir_path_list = convert_dir_path_to_list(cwd);
	vector<string> new_dir_path_list = convert_dir_path_to_list(to_dirctory);

	vector<string> new_dir = prepare_new_dir(cwd_dir_path_list, new_dir_path_list);
        vector<string>::iterator it;

        for(it = new_dir.begin(); it != new_dir.end(); it++) {
	        cout<<*it;
        }
        cout<<"\n";

    } catch(string exp) { // Default exception
        cout<<"mycd: "<<exp<<"\n";
    }
    return 0;
}

File: linux_cd_cmd_simulation/go/mycd.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

func remove_index_from_slice_of_strings(slice []string, index int) []string {
	return append(slice[:index], slice[index+1:]...)
}

func panic_recovery(err error) {
	if err := recover(); err != nil { //catch
		fmt.Println("mycd: ", err)
		os.Exit(1)
	}
}

func is_alphanum(word string) bool {
	return regexp.MustCompile(`^[a-zA-Z0-9]*$`).MatchString(word)
}

func remove_empty_index_in_slice(slice []string, from_index int) []string {
	for i := from_index; i < len(slice); {
		if strings.Compare(slice[i], "") == 0 {
			slice = remove_index_from_slice_of_strings(slice, i)
		} else {
			break
		}
	}
	return slice
}

func validate_and_convert_dir_path_to_slice(dir_path string) []string {

	dir_path_list := strings.Split(dir_path, "/")

	if strings.Compare(dir_path_list[0], "") == 0 {
		dir_path_list = remove_empty_index_in_slice(dir_path_list, 1)
	}

	for i, dir_name := range dir_path_list {
		// Validate if dir name is alphanumeric or not
		if !(is_alphanum(dir_name) ||
			(strings.Compare(dir_name, "..") == 0) ||
			(strings.Compare(dir_name, ".") == 0)) {
			panic(dir_name + ": No such file or directory")
		}
		// Remove extra backslashes if persists (In strings.Split with "/"
		// converts backslashes to empty string)
		if strings.Compare(dir_name, "") == 0 {
			if i == 0 {
				// ignore first backslash if exists
				continue
			}
			dir_path_list = remove_empty_index_in_slice(dir_path_list, i)
		}
	}

	return dir_path_list
}

func back_to_the_child_path(cwd []string) []string {
	return append(cwd[:len(cwd)-1], cwd[len(cwd):]...)
}

func prepare_new_dir(cwd_dir_path_list, new_dir_path_list []string) string {

	new_dir := "/"

	if strings.Compare(new_dir_path_list[0], "") == 0 {
		// If target path starts with "/". Checking with empty string because In strings.Split with "/"
		// "/" converts backslash to empty string in slice of strings
		if len(new_dir_path_list) == 1 && strings.Compare(new_dir_path_list[0], "") == 0 {
			// If target has only backslash "/"
			return new_dir
		}
		cwd_dir_path_list = nil
		cwd_dir_path_list = append(cwd_dir_path_list, "/")
	}

	for i := 0; i < len(new_dir_path_list); i++ {
		if strings.Compare(new_dir_path_list[i], "") == 0 ||
			strings.Compare(new_dir_path_list[i], ".") == 0 {
			continue
		}
		if strings.Compare(new_dir_path_list[i], "..") == 0 {
			if len(cwd_dir_path_list) > 1 {
				cwd_dir_path_list = back_to_the_child_path(cwd_dir_path_list)
			}
		} else {
			cwd_dir_path_list = append(cwd_dir_path_list, new_dir_path_list[i])
		}
	}
	if len(cwd_dir_path_list) == 1 && strings.Compare(cwd_dir_path_list[0], "") == 0 {
		return new_dir
	} else {
		new_dir = strings.Join(cwd_dir_path_list, "/")
		if (len(new_dir) > 1) &&
			strings.Compare(string(new_dir[0]), "/") == 0 &&
			strings.Compare(string(new_dir[1]), "/") == 0 {
			new_dir = new_dir[1:]
		}
		return new_dir
	}
}

func main() {
	defer panic_recovery(nil)

	if len(os.Args) < 3 {
		panic("too few arguments")
	} else if len(os.Args) > 3 {
		panic("too many arguments")
	}

	cwd := os.Args[1]
	to_dirctory := os.Args[2]

	cwd_dir_path_list := validate_and_convert_dir_path_to_slice(cwd)
	new_dir_path_list := validate_and_convert_dir_path_to_slice(to_dirctory)

	new_dir := prepare_new_dir(cwd_dir_path_list, new_dir_path_list)

	fmt.Println(new_dir)
}
